src/main/java/io/cryostat/discovery/DiscoveryStorage.java

```diff
@@ -228,10 +235,31 @@ private Map<String, String> mergeLabels(
 
     public List<? extends AbstractNode> update(
             UUID id, Collection<? extends AbstractNode> children) {
+        var updatedChildren = new ArrayList<AbstractNode>();
+        for (AbstractNode child : Objects.requireNonNull(children, "children")) {
```

Comment:
The subtree being pushed here may not be "flat", with the collection being a direct list of `TargetNodes`. It could be nested like in a k8s environment where it does something like namespace -> deployment -> replica -> endpoints -> serviceref. I think you should use `findLeavesFrom()` on each of the `children` here, then merge all of those results into a single `Collection` and then iterate over those and do the JVM ID retrieval.

Reply:
Right, forgot about that. Thanks!

Reply:
Just realized something, I would have to mutate the actual ServiceRef objects themselves such that the structure of the children isn't messed up when calling `dao.update()`, is that okay to do or is there a better way?

Reply:
You mean to remove the invalid ones from the tree, where the JVM ID retrieval fails?

Reply:
I mean I think if I did something like this
```
    public List<? extends AbstractNode> update(
            UUID id, Collection<? extends AbstractNode> children) {
        var updatedChildren = new HashSet<AbstractNode>();
        for (AbstractNode child : findAllLeavesFrom(Objects.requireNonNull(children, "children"))) {
            if (child instanceof TargetNode) {
                ServiceRef ref = ((TargetNode) child).getTarget();
                 ref = jvmIdHelper.get().resolveId(ref);
                child = new TargetNode(child.getNodeType(), ref, child.getLabels());
            }
            updatedChildren.add(child);
        }
...
        plugin = dao.update(id, Objects.requireNonNull(updatedChildren));
```
I don't think it would work since now the updateChildren is flattened again, so if I wanted to update the jvmIds along the way, I would have to, iterate the children leaves again but **not** make a new Collection and modify the leaves' ServiceRef objects directly with something like TargetNode.setJvmId(). Then this way we can just update the dao as normal with children
`plugin = dao.update(id, children);` since the structure wasn't flattened. Another way to do would I guess to reconstruct the tree structure as we are iterating so that mutations are not needed.